File: challenges/leetcode/golang/searchInsertPosition.go
package main

import "log"

func searchInsert(nums []int, target int) int {
	hi := len(nums) - 1
	lo := 0

	for {
		diff := hi - lo
		if diff <= 1 {
			break
		}

		mid := diff / 2

		if nums[mid] > target {
		}
	}

	return 0
}

func main() {
	log.Print(searchInsert([]int{1, 3, 5, 6}, 5))
}

File: typescript/src/leetcode/searchInsertPosition.ts
//Non recursive zero copy modified bin search
//O(log(n))

function searchInsert(nums: number[], target: number): number {
    let lo = 0
    let hi = nums.length - 1
    let mid = Math.floor(hi / 2)

    while (hi - lo > 0) {
        if (nums[mid] == target) return mid
        if (nums[mid] > target) {
            hi = mid
            mid = Math.floor(hi / 2)
            continue
        }
        lo = mid + 1
        mid = Math.floor((hi - lo) / 2) + lo
    }

    if (nums[lo] >= target) return lo
    return lo + 1
};

export default searchInsert
